### PR TITLE
azure application identity and permission working

### PR DESCRIPTION
**Pull request by chrisghill**

### CODE REVIEW

File: internal/cloud/azure/application_identity.go
Comment (chrisghill) [inline internal/cloud/azure/application_identity.go:52]: ClientID is an overloaded term. Azure applications can have a `client_id` (different than the actual ID 👍 ), but permissions can't be attached to applications.  Only to service principles, which are "owned" by applications (but somehow unviewable in the console???).

Its messy, but I changed the name here to be explicit that this is a **Service Principal** ID, not the application `client_id`.

File: internal/cloud/azure/application_permission.go
Comment (chrisghill) [inline internal/cloud/azure/application_permission.go:26]: This is a really dumb headache. Authorization in the new client library is a lot easier, but this is what we are stuck with on the old client.  Don't waste brain cells reviewing this, its basically copied from [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/accdeebe6072390c531b88f8e322b80acceecfb8/internal/clients/builder.go#L40-L95).

File: internal/cloud/azure/application_permission.go
Comment (chrisghill) [inline internal/cloud/azure/application_permission.go:102]: So this is a strange Azure-ism...Roles don't seem to be "singleton".  They vary based off of the "scope".  So we have to do a lookup to get the Role ID by looking up the RoleName (ex `Storage Account Contributor`) with a scope.

Technically we *could* change up our security block and have the bundle output the **fully scoped** ID of the role (`/subscriptions/sdkfjldf/Microsoft.something/lskdjfskldj/Roles/sldkfsldkjf/Whatever`) and we might be able to skip this step, but I don't know if that would play nicely with other clients.  I pulled all this code from their own terraform provider, so it seems they kinda expect to jump through these hoops on Role lookup.

File: internal/cloud/azure/application_permission.go
Comment (chrisghill) [inline internal/cloud/azure/application_permission.go:131]: Calling a wrapper function for create so we can implement retries

File: internal/cloud/azure/application_permission.go
Comment (chrisghill) [inline internal/cloud/azure/application_permission.go:151]: At least `delete` is easy...

File: internal/cloud/azure/application_permission.go
Comment (chrisghill) [inline internal/cloud/azure/application_permission.go:186]: Don't rack your brain too hard here, again its [mostly copied](https://github.com/hashicorp/terraform-provider-azurerm/blob/accdeebe6072390c531b88f8e322b80acceecfb8/internal/services/authorization/role_assignment_resource.go#L317-L327) with our own retry wrapper.

The reason for this is that the service principal `create` can return very quickly, but that doesn't mean you can attach roles yet.  I tried backing this "polling" back to the SP creation itself, but of course that doesn't work...reads and lookups work just fine.  Just these role attachments (and who knows what else) will fail until after some arbitrary time when its "ready".

Azure does the same thing in their provider...the polling occurs downstream instead of the SP creation blocking until its "ready". Its probably just not easy to determine via the API.

File: internal/cloud/azure/utils.go
Comment (chrisghill) [inline internal/cloud/azure/utils.go:1]: [COPY PASTA](https://github.com/hashicorp/terraform-provider-azurerm/blob/accdeebe6072390c531b88f8e322b80acceecfb8/utils/response.go#L42)

File: internal/cloud/azure/application_identity.go
Comment (chrisghill) [inline internal/cloud/azure/application_identity.go:82]: Notice the switch-a-roo magic we are doing here.  The `id` of this resource is the **Service Principal ID**, not the Application ID (or client id).  This is so the `application_permission` for azure can work just like GCP and AWS and just do:

```
resource "mdxc_application_permission" "example" {
  application_identity_id = mdxc_application_identity.example.id
  permission = {...}
}
```

instead of:

```
resource "mdxc_application_permission" "example" {
  application_identity_id = mdxc_application_identity.example.azure_application_identity.service_principal_id
  permission = {...}
}
```

However, this does put us in a weird situation where the application *might* have been made in a previous run, but the SP creation failed, so the ID doesn't get updated or stored in state. The application will be orphaned in this scenario.  Not sure how to address this other than deleting the Application resource if the SP creation fails (which I'm not doing yet).

File: internal/cloud/azure/application_permission.go
Comment (chrisghill) [inline internal/cloud/azure/application_permission.go:170]: Mostly [copy pasta](https://github.com/hashicorp/terraform-provider-azurerm/blob/accdeebe6072390c531b88f8e322b80acceecfb8/internal/services/authorization/role_assignment_resource.go#L360-L380)

File: internal/cloud/azure/application_identity.go
Comment (coryodaniel) [reply internal/cloud/azure/application_identity.go:90]: Id prefer not to expose a cloud specific boolean, is there enough information to know its going to be running on kubernetes?

File: internal/cloud/azure/application_identity.go
Comment (coryodaniel) [reply internal/cloud/azure/application_identity.go:52]: How do you access the id and secret from the terraform resource?

Comment (chrisghill) [reply internal/cloud/azure/application_identity.go:52]: The resource `id` field will be the Service Principal ID, but both fields are directly accessible in the azure-specific output block.  You can see I updated the name of them here: https://github.com/massdriver-cloud/terraform-provider-mdxc/pull/25/files#diff-08d112e976e2479674666bd6acea8430fd217225c39ce5186b3e1a4788d52ad1L86-L92

File: internal/cloud/azure/application_identity.go
Comment (chrisghill) [reply internal/cloud/azure/application_identity.go:90]: > Id prefer not to expose a cloud specific boolean, is there enough information to know its going to be running on kubernetes?

Agreed on not exposing the boolean, which is why it's hardcoded to creation for now.  There currently isn't enough information to determine if its Kubernetes or not. I figure we should move a `kubernetes` block to the top level (@WillBeebe [added a `kubernetes` specific block to the GCP specific inputs](https://github.com/massdriver-cloud/terraform-provider-mdxc/blob/main/internal/provider/resource_application_identity.go#L50-L63).

Both GCP and Azure have this problem of needed to take different actions if its a kubernetes cluster, though Azure hopefully only until AKS workload identity is supported.  In AWS this is handled via the `assume_role_policy` which we are currently doing in terraform locals instead of the provider (though we could move it into the provider).

File: internal/cloud/azure/application_identity.go
Comment (WillBeebe) [reply internal/cloud/azure/application_identity.go:82]: @mclacore for context. This PR creates most of the resources in your file here: https://github.com/massdriver-cloud/terraform-modules/pull/33/files#r960073173

And that's on purpose. :) Chris has smashed this, and you the app side. 👏  Once this is merged that file will look a lot like my comment.

**Review by coryodaniel — APPROVED** [review]

**Comment by coryodaniel** [comment]
Install precommit and run it !

File: internal/cloud/azure/application_identity.go
Comment (WillBeebe) [reply internal/cloud/azure/application_identity.go:90]: Stretched the legs of this PR w/ Michael this afternoon. We ended up needing the service principal password to pull from the Azure container repo. 

I'd say we merge this, the terraform module, and the app template. Michael has the module and template in really good shape and they're almost ready for review. 

After that pass, we can do k8s mdx and I can add a `kubernetes` block here if we need it. LMK what you think.

File: internal/cloud/azure/application_permission.go
Comment (WillBeebe) [reply internal/cloud/azure/application_permission.go:102]: great context

Comment (WillBeebe) [reply internal/cloud/azure/application_permission.go:102]: I think it's a good idea to mimic their provider and leave the security blocks as-is for now.

**Review by WillBeebe — APPROVED** [review]
LGTM. Michael made an Azure function with this today and we didn't have any issues. We didn't test the security block stuff yet, but I think we can merge this and follow up if needed.

**Pull request merged**
